lib/elixir_sense.ex

```diff
@@ -332,26 +332,31 @@ defmodule ElixirSense do
       ...>   C.func()
       ...> end
       ...> '''
-      iex> ElixirSense.references(code, 3, 6) |> Enum.take(2)
+      ...> trace = %{
+      ...>   {ElixirSense.Providers.ReferencesTest.Modules.Callee1, :func, 0} => %{
+      ...>     callee: {ElixirSense.Providers.ReferencesTest.Modules.Callee1, :func, 0},
+      ...>     file: "test/support/modules_with_references.ex",
+      ...>     line: 36
+      ...>   }
+      ...> }
+      iex> ElixirSense.references(code, 3, 6, trace)
       [
         %{
           uri: "test/support/modules_with_references.ex",
           range: %{
             start: %{line: 36, column: 60},
             end: %{line: 36, column: 64}
           }
-        },
-        %{
-          uri: "test/support/modules_with_references.ex",
-          range: %{
-            start: %{line: 65, column: 16},
-            end: %{line: 65, column: 20}
-          }
         }
       ]
   """
-  @spec references(String.t(), pos_integer, pos_integer) :: [References.reference_info()]
-  def references(code, line, column) do
+  @spec references(
+          String.t(),
+          pos_integer,
+          pos_integer,
+          ElixirSense.Core.References.Tracer.call_trace_t() | nil
+        ) :: [References.reference_info()]
+  def references(code, line, column, trace \\ nil) do
```

Comment:
I think that requiring passing in a trace (that was generated by `ElixirSense.Core.References.Tracer` which isn't marked private, although it probably should be) to a public api could be improved because now callers of `ElixirSense.references/4` need to call a potentially private api.

Instead perhaps we should collect and store the compiler trace within ElixirSense itself, although that would require ElixirSense to have it's own supervision hierarchy. Any thoughts on the tradeoffs involved?

Reply:
Obviously the tracer API needs to be public.

Regarding setting up a supervision hierarchy in ElixirSense, it would be a major breaking change for the users. ElixirSense is currently a stateless library. For it to work It requires the client to compile and load all the modules, as it is in the case of elixir-LS. In my opinion it is easier to setup and manage the tracer process there.

Reply:
Well changing ElixirSense to have it's own supervision hierarchy should generally be transparent to end-users, although I can see an argument that managing the tracer process would be easier from ElixirLS :thinking:

Reply:
I agree with @axelson, ElixirSense could have its own supervision tree that end-users would not have to know about, especially since ElixirSense is largely obscured behind ElixirLS. Supervision trees are simple to set up, and I could make a PR to add just that to ElixirSense (that I imagine would just supervise the TCP server at this stage).

Alternatively, could we use `:ets` for storage of the traces and not have to manually manage supervision?

Wandering perhaps too far into the philosophical; Is there a reason other than historical for why ElixirSense and ElixirLS are separate codebases? Wouldn't changes like these be easier to think about if they were unified?